collector/zfs_linux.go

```diff
@@ -0,0 +1,83 @@
+package collector
+
+// +build linux
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/common/log"
+)
+
+const zfsArcstatsProcpath = "spl/kstat/zfs/arcstats"
+
+func zfsInitialize() error {
+	path := procFilePath(zfsArcstatsProcpath)
+	fd, err := os.Open(path)
+	if err != nil {
+		log.Errorf("Cannot open ZFS arcstats logpath: %s", path)
+		return err
+	}
+	return fd.Close()
+}
+
+func (p *zfsMetricProvider) PrepareUpdate() (err error) {
+
+	err = p.prepareUpdateArcstats(zfsArcstatsProcpath)
+	if err != nil {
+		return
+	}
+	return nil
+}
+
+func (p *zfsMetricProvider) handleFetchedMetricCacheMiss(s zfsSysctl) (value zfsMetricValue, err error) {
+
+	// all values are fetched in PrepareUpdate()
+	return zfsErrorValue, fmt.Errorf("sysctl '%s' found")
+}
+
+func (p *zfsMetricProvider) prepareUpdateArcstats(zfsArcstatsProcpath string) (err error) {
+
+	file, err := os.Open(procFilePath(zfsArcstatsProcpath))
+	if err != nil {
+		log.Errorf("Cannot open ZFS arcstats procfs file for reading. " +
```

Comment:
We should probably enable this collector by default, and then not treat this as an error. I don't see any danger in having it running on systems not using zfs.